BackwardSyncException: Did not receive a header for hash {}0x0000000000000000000000000000000000000000000000000000000000000000

<!-- Have you done the following? -->
<!--   * read the Code of Conduct? By filing an Issue, you are expected to -->  
<!--     comply with it, including treating everyone with respect: -->
<!--     https://github.com/hyperledger/besu/blob/main/CODE_OF_CONDUCT.md -->
<!--   * Reproduced the issue in the latest version of the software -->
<!--   * Read the debugging docs: https://besu.hyperledger.org/en/stable/HowTo/Monitor/Logging/ -->
<!--   * Duplicate Issue check:  https://github.com/search?q=+is%3Aissue+repo%3Ahyperledger/Besu -->
<!-- Note:  Not all sections will apply to all issue types. -->

### Description

Running Hive test **Two Block PoW Re-org to Higher-Height Chain**, that does this:

> Client 1 starts with chain G -> A, Client 2 starts with chain G -> B -> C.  
> Blocks A and C reach TTD, but block C has higher height than A.  
> ForkchoiceUpdated is sent to Client 1 with A as Head.  
> ForkchoiceUpdated is sent to Client 1 with C as Head.  
> Verification is made that Client 1 Re-orgs to chain G -> B -> C.  

 this exception is produced

```
2022-05-19 17:17:26.505+00:00 | vert.x-worker-thread-0 | DEBUG | MergeCoordinator | appending block hash 0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4 to backward sync
2022-05-19 17:17:26.506+00:00 | vert.x-worker-thread-0 | INFO  | BackwardSyncContext | Starting new backward sync towards a pivot 0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4
2022-05-19 17:17:26.517+00:00 | EthScheduler-Workers-0 | DEBUG | RetryingGetBlockFromPeersTask | Getting block -1 (Optional[0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4]) from peer Optional[Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false], attempt 1
2022-05-19 17:17:26.518+00:00 | EthScheduler-Workers-0 | DEBUG | GetBlockFromPeerTask | Downloading block -1 (Optional[0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4]) from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false.
2022-05-19 17:17:26.526+00:00 | EthScheduler-Workers-0 | DEBUG | GetHeadersFromPeerByHashTask | Requesting 1 headers from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false.
2022-05-19 17:17:26.529+00:00 | EthScheduler-Workers-0 | TRACE | AbstractPeerConnection | Writing RawMessage{code=3, data=0xe601e4a02d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4018080} to PeerInfo{version=5, clientId='besu/v22.4.1-dev/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, snap/1], port=30303, nodeId=0x0787b45542bbef79f2f1f2363e0ca21a26c0b8307ec5f37fe825990f3535ef2d742234cbe6a5f2c6b0e974264ba3430e002455510c949004251a3508c7dcb3d7} via protocol eth/66
2022-05-19 17:17:26.533+00:00 | nioEventLoopGroup-3-2 | TRACE | EthProtocolManager | Process message eth/66, 4
2022-05-19 17:17:26.534+00:00 | nioEventLoopGroup-3-2 | DEBUG | AbstractGetHeadersFromPeerTask | Received 1 of 1 headers requested from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false
2022-05-19 17:17:26.539+00:00 | nioEventLoopGroup-3-2 | DEBUG | GetBodiesFromPeerTask | Requesting 1 bodies from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false.
2022-05-19 17:17:26.543+00:00 | nioEventLoopGroup-3-2 | TRACE | AbstractPeerConnection | Writing RawMessage{code=5, data=0xe380e1a02d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4} to PeerInfo{version=5, clientId='besu/v22.4.1-dev/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, snap/1], port=30303, nodeId=0x0787b45542bbef79f2f1f2363e0ca21a26c0b8307ec5f37fe825990f3535ef2d742234cbe6a5f2c6b0e974264ba3430e002455510c949004251a3508c7dcb3d7} via protocol eth/66
2022-05-19 17:17:26.557+00:00 | nioEventLoopGroup-3-2 | TRACE | EthProtocolManager | Process message eth/66, 6
2022-05-19 17:17:26.560+00:00 | nioEventLoopGroup-3-2 | DEBUG | GetBlockFromPeerTask | Successfully downloaded block -1 (Optional[0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4]) from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false.
2022-05-19 17:17:26.560+00:00 | nioEventLoopGroup-3-2 | DEBUG | SyncStepStep | Appending block 2(0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4)
2022-05-19 17:17:26.569+00:00 | nioEventLoopGroup-3-2 | INFO  | BackwardSyncStep | lastHash -> 0x548b3a20d40c9b29fe6bf89b86bc3e1d52c1fadd60d1590627f486b6cd56635c
2022-05-19 17:17:26.570+00:00 | nioEventLoopGroup-3-2 | DEBUG | BackwardSyncStep | Requesting header for hash 0x548b3a20d40c9b29fe6bf89b86bc3e1d52c1fadd60d1590627f486b6cd56635c
2022-05-19 17:17:26.577+00:00 | EthScheduler-Workers-1 | DEBUG | GetHeadersFromPeerByHashTask | Requesting 200 headers from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false.
2022-05-19 17:17:26.578+00:00 | EthScheduler-Workers-1 | TRACE | AbstractPeerConnection | Writing RawMessage{code=3, data=0xe702e5a0548b3a20d40c9b29fe6bf89b86bc3e1d52c1fadd60d1590627f486b6cd56635c81c88001} to PeerInfo{version=5, clientId='besu/v22.4.1-dev/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, snap/1], port=30303, nodeId=0x0787b45542bbef79f2f1f2363e0ca21a26c0b8307ec5f37fe825990f3535ef2d742234cbe6a5f2c6b0e974264ba3430e002455510c949004251a3508c7dcb3d7} via protocol eth/66
2022-05-19 17:17:26.584+00:00 | nioEventLoopGroup-3-2 | TRACE | EthProtocolManager | Process message eth/66, 4
2022-05-19 17:17:26.594+00:00 | nioEventLoopGroup-3-2 | DEBUG | AbstractGetHeadersFromPeerTask | Received 2 of 200 headers requested from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false
2022-05-19 17:17:26.596+00:00 | nioEventLoopGroup-3-2 | DEBUG | BackwardSyncStep | Got headers 1 -> 0
2022-05-19 17:17:26.599+00:00 | nioEventLoopGroup-3-2 | DEBUG | BackwardChain | Added header 0x548b3a20d40c9b29fe6bf89b86bc3e1d52c1fadd60d1590627f486b6cd56635c on height 1 to backward chain led by pivot 0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4 on height 2
2022-05-19 17:17:26.600+00:00 | nioEventLoopGroup-3-2 | DEBUG | BackwardChain | Added header 0x67ead97eb79b47a1638659942384143f36ed44275d4182799875ab5a87324055 on height 0 to backward chain led by pivot 0x2d4f65a26a8d5ca4f8000010b4741652decb89c8ccf19119b5d9f58846efbbd4 on height 1
2022-05-19 17:17:26.602+00:00 | nioEventLoopGroup-3-2 | INFO  | BackwardSyncStep | Saved headers 1 -> 0 (head: 1)
2022-05-19 17:17:26.605+00:00 | nioEventLoopGroup-3-2 | INFO  | BackwardSyncStep | lastHash -> 0x0000000000000000000000000000000000000000000000000000000000000000
2022-05-19 17:17:26.605+00:00 | nioEventLoopGroup-3-2 | DEBUG | BackwardSyncStep | Requesting header for hash 0x0000000000000000000000000000000000000000000000000000000000000000
2022-05-19 17:17:26.616+00:00 | EthScheduler-Workers-2 | DEBUG | GetHeadersFromPeerByHashTask | Requesting 200 headers from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false.
2022-05-19 17:17:26.617+00:00 | EthScheduler-Workers-2 | TRACE | AbstractPeerConnection | Writing RawMessage{code=3, data=0xe703e5a0000000000000000000000000000000000000000000000000000000000000000081c88001} to PeerInfo{version=5, clientId='besu/v22.4.1-dev/linux-x86_64/openjdk-java-11', capabilities=[eth/62, eth/63, eth/64, eth/65, eth/66, snap/1], port=30303, nodeId=0x0787b45542bbef79f2f1f2363e0ca21a26c0b8307ec5f37fe825990f3535ef2d742234cbe6a5f2c6b0e974264ba3430e002455510c949004251a3508c7dcb3d7} via protocol eth/66
2022-05-19 17:17:26.623+00:00 | nioEventLoopGroup-3-2 | TRACE | EthProtocolManager | Process message eth/66, 4
2022-05-19 17:17:26.624+00:00 | nioEventLoopGroup-3-2 | DEBUG | AbstractGetHeadersFromPeerTask | headers.isEmpty. Peer: Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false
2022-05-19 17:17:26.625+00:00 | vert.x-worker-thread-12 | DEBUG | JsonRpcExecutor | JSON-RPC request -> eth_getBlockByNumber [latest, false]
2022-05-19 17:17:26.637+00:00 | nioEventLoopGroup-3-2 | DEBUG | EthPeer | Received useless response for headers from peer Peer 0x0787b45542bbef79f2... PeerReputation 100, validated? true, disconnected? false
2022-05-19 17:17:26.656+00:00 | nioEventLoopGroup-3-2 | WARN  | BackwardSyncContext | there was an uncaught exception during backward sync
java.util.concurrent.CompletionException: org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardSyncException: Did not receive a header for hash {}0x0000000000000000000000000000000000000000000000000000000000000000
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
        at org.hyperledger.besu.util.FutureUtils.lambda$propagateResult$1(FutureUtils.java:74)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
        at org.hyperledger.besu.ethereum.eth.manager.task.RetryingGetHeadersEndingAtFromPeerByHashTask.lambda$executePeerTask$0(RetryingGetHeadersEndingAtFromPeerByHashTask.java:88)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
        at org.hyperledger.besu.ethereum.eth.manager.task.AbstractPeerRequestTask.lambda$executeTask$2(AbstractPeerRequestTask.java:83)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
        at java.base/java.util.Optional.ifPresent(Optional.java:183)
        at org.hyperledger.besu.ethereum.eth.manager.task.AbstractPeerRequestTask.handleMessage(AbstractPeerRequestTask.java:104)
        at org.hyperledger.besu.ethereum.eth.manager.task.AbstractPeerRequestTask.lambda$executeTask$0(AbstractPeerRequestTask.java:67)
        at org.hyperledger.besu.ethereum.eth.manager.RequestManager$ResponseStream.dispatchBufferedResponses(RequestManager.java:191)
        at org.hyperledger.besu.ethereum.eth.manager.RequestManager$ResponseStream.close(RequestManager.java:170)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.hyperledger.besu.ethereum.eth.manager.RequestManager.closeOutstandingStreams(RequestManager.java:99)
        at org.hyperledger.besu.ethereum.eth.manager.RequestManager.dispatchResponse(RequestManager.java:83)
        at org.hyperledger.besu.ethereum.eth.manager.EthPeer.lambda$dispatch$7(EthPeer.java:367)
        at java.base/java.util.Optional.ifPresentOrElse(Optional.java:201)
        at org.hyperledger.besu.ethereum.eth.manager.EthPeer.dispatch(EthPeer.java:366)
        at org.hyperledger.besu.ethereum.eth.manager.EthPeers.dispatchMessage(EthPeers.java:139)
        at org.hyperledger.besu.ethereum.eth.manager.EthProtocolManager.processMessage(EthProtocolManager.java:275)
        at org.hyperledger.besu.ethereum.p2p.network.NetworkRunner.lambda$setupHandlers$0(NetworkRunner.java:135)
        at org.hyperledger.besu.ethereum.p2p.rlpx.connections.PeerConnectionEvents.lambda$dispatchMessage$1(PeerConnectionEvents.java:64)
        at org.hyperledger.besu.util.Subscribers.lambda$forEach$0(Subscribers.java:112)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.hyperledger.besu.util.Subscribers.forEach(Subscribers.java:109)
        at org.hyperledger.besu.ethereum.p2p.rlpx.connections.PeerConnectionEvents.dispatchMessage(PeerConnectionEvents.java:64)
        at org.hyperledger.besu.ethereum.p2p.rlpx.connections.netty.ApiHandler.channelRead0(ApiHandler.java:99)
        at org.hyperledger.besu.ethereum.p2p.rlpx.connections.netty.ApiHandler.channelRead0(ApiHandler.java:33)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:327)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:299)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:722)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:658)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:584)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardSyncException: Did not receive a header for hash {}0x0000000000000000000000000000000000000000000000000000000000000000
        at org.hyperledger.besu.ethereum.eth.sync.backwardsync.BackwardSyncStep.lambda$requestHeaders$1(BackwardSyncStep.java:85)
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
        ... 63 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,10 @@
 # Changelog
 
+## 22.4.2
+
+### Bug Fixes
+- Stop backward sync if genesis block has been reached [#3869](https://github.com/hyperledger/besu/pull/3869)
+
 ## 22.4.1
 
 ### Additions and Improvements

--- a/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardSyncContext.java
+++ b/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardSyncContext.java
@@ -214,12 +214,27 @@ public class BackwardSyncContext {
           firstAncestorHeader.getHash());
     }
 
-    if (getProtocolContext().getBlockchain().contains(firstAncestorHeader.getParentHash())
-        || getProtocolContext()
-            .getBlockchain()
-            .getBlockByHash(firstAncestorHeader.getHash())
-            .filter(block -> block.getHeader().getNumber() == 0)
-            .isPresent()) {
+    if (firstAncestorHeader.getNumber() == 0) {
+      final BlockHeader genesisBlockHeader =
+          getProtocolContext()
+              .getBlockchain()
+              .getBlockHeader(0)
+              .orElseThrow(
+                  () -> new IllegalStateException("Really we do not have the genesis block?"));
+
+      if (genesisBlockHeader.getHash().equals(firstAncestorHeader.getHash())) {
+        LOG.info("Backward sync reached genesis, starting Forward sync");
+        return executeForwardAsync(genesisBlockHeader);
+      } else {
+        return CompletableFuture.failedFuture(
+            new BackwardSyncException(
+                String.format(
+                    "Backward sync reached genesis, but ancestor header hash %s does not match our genesis header hash %s",
+                    firstAncestorHeader.getHash(), genesisBlockHeader.getHash())));
+      }
+    }
+
+    if (getProtocolContext().getBlockchain().contains(firstAncestorHeader.getParentHash())) {
       return executeForwardAsync(firstAncestorHeader);
     }
     return executeBackwardAsync(firstAncestorHeader);

--- a/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardSyncContext.java
+++ b/ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardSyncContext.java
@@ -195,24 +195,34 @@ public class BackwardSyncContext {
     if (!isReady()) {
       return waitForTTD().thenCompose(this::executeNextStep);
     }
-    final Optional<BlockHeader> firstAncestorHeader = backwardChain.getFirstAncestorHeader();
-    if (firstAncestorHeader.isEmpty()) {
+
+    final Optional<BlockHeader> maybeFirstAncestorHeader = backwardChain.getFirstAncestorHeader();
+    if (maybeFirstAncestorHeader.isEmpty()) {
       LOG.info("The Backward sync is done...");
       return CompletableFuture.completedFuture(null);
     }
+
+    final BlockHeader firstAncestorHeader = maybeFirstAncestorHeader.get();
+
     if (getProtocolContext().getBlockchain().getChainHead().getHeight()
-        > firstAncestorHeader.get().getNumber() - 1) {
+        > firstAncestorHeader.getNumber() - 1) {
       LOG.info(
-          "Backward reached bellow previous head {}({}) : {} ({})",
+          "Backward reached below previous head {}({}) : {} ({})",
           getProtocolContext().getBlockchain().getChainHead().getHeight(),
           getProtocolContext().getBlockchain().getChainHead().getHash().toHexString(),
-          firstAncestorHeader.get().getNumber(),
-          firstAncestorHeader.get().getHash());
+          firstAncestorHeader.getNumber(),
+          firstAncestorHeader.getHash());
     }
-    if (getProtocolContext().getBlockchain().contains(firstAncestorHeader.get().getParentHash())) {
-      return executeForwardAsync(firstAncestorHeader.get());
+
+    if (getProtocolContext().getBlockchain().contains(firstAncestorHeader.getParentHash())
+        || getProtocolContext()
+            .getBlockchain()
+            .getBlockByHash(firstAncestorHeader.getHash())
+            .filter(block -> block.getHeader().getNumber() == 0)
+            .isPresent()) {
+      return executeForwardAsync(firstAncestorHeader);
     }
-    return executeBackwardAsync(firstAncestorHeader.get());
+    return executeBackwardAsync(firstAncestorHeader);
   }
 
   private CompletableFuture<Void> executeSyncStep(final Hash hash) {

--- a/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardSyncContextTest.java
+++ b/ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/backwardsync/BackwardSyncContextTest.java
@@ -51,6 +51,7 @@ import org.hyperledger.besu.plugin.services.BesuEvents;
 import org.hyperledger.besu.plugin.services.MetricsSystem;
 import org.hyperledger.besu.services.kvstore.InMemoryKeyValueStorage;
 
+import java.nio.charset.StandardCharsets;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
@@ -58,6 +59,7 @@ import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
 import javax.annotation.Nonnull;
 
+import org.apache.tuweni.bytes.Bytes;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.runner.RunWith;
@@ -80,6 +82,7 @@ public class BackwardSyncContextTest {
   private MutableBlockchain remoteBlockchain;
   private RespondingEthPeer peer;
   private MutableBlockchain localBlockchain;
+  private Block genesisBlock;
 
   @Spy
   private ProtocolSchedule protocolSchedule =
@@ -101,7 +104,7 @@ public class BackwardSyncContextTest {
   public void setup() {
     when(mockProtocolSpec.getBlockValidator()).thenReturn(blockValidator);
     when(protocolSchedule.getByBlockNumber(anyLong())).thenReturn(mockProtocolSpec);
-    Block genesisBlock = blockDataGenerator.genesisBlock();
+    genesisBlock = blockDataGenerator.genesisBlock();
     remoteBlockchain = createInMemoryBlockchain(genesisBlock);
     localBlockchain = createInMemoryBlockchain(genesisBlock);
 
@@ -277,9 +280,50 @@ public class BackwardSyncContextTest {
     verify(context).executeBackwardAsync(any());
   }
 
+  @Test
+  public void shouldStartForwardSyncIfGenesisIsReached() {
+    final MutableBlockchain otherLocalBlockchain = createForkedBlockchain(genesisBlock);
+    when(protocolContext.getBlockchain()).thenReturn(otherLocalBlockchain);
+
+    createBackwardChain(0, 1);
+    doReturn(CompletableFuture.completedFuture(null)).when(context).executeForwardAsync(any());
+
+    context.executeNextStep(null);
+    verify(context).executeForwardAsync(any());
+  }
+
+  @Test
+  public void shouldFailIfADifferentGenesisIsReached() {
+    final BlockDataGenerator.BlockOptions otherGenesisOptions =
+        new BlockDataGenerator.BlockOptions()
+            .setBlockNumber(0)
+            .setExtraData(Bytes.of("Other genesis".getBytes(StandardCharsets.UTF_8)));
+    final Block otherGenesis = blockDataGenerator.genesisBlock(otherGenesisOptions);
+    final MutableBlockchain otherLocalBlockchain = createForkedBlockchain(otherGenesis);
+    when(protocolContext.getBlockchain()).thenReturn(otherLocalBlockchain);
+
+    createBackwardChain(0, 1);
+    CompletableFuture<Void> res = context.executeNextStep(null);
+    assertThat(res.isCompletedExceptionally()).isTrue();
+  }
+
   private void createBackwardChain(final int from, final int until) {
     for (int i = until; i > from; --i) {
       backwardChain.prependAncestorsHeader(getBlockByNumber(i - 1).getHeader());
     }
+  }
+
+  private MutableBlockchain createForkedBlockchain(final Block genesisBlock) {
+    final MutableBlockchain otherLocalBlockchain = createInMemoryBlockchain(genesisBlock);
+    final BlockDataGenerator.BlockOptions options =
+        new BlockDataGenerator.BlockOptions()
+            .setBlockNumber(1)
+            .setParentHash(genesisBlock.getHash())
+            .setExtraData(Bytes.of("Fork of block 1".getBytes(StandardCharsets.UTF_8)));
+    final Block block = blockDataGenerator.block(options);
+    final List<TransactionReceipt> receipts = blockDataGenerator.receipts(block);
+
+    otherLocalBlockchain.appendBlock(block, receipts);
+    return otherLocalBlockchain;
   }
 }

